File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-7.java
class Solution {
    //1 2 3 4 5 6
    //6 1 2 3 4 5
    //5 6 1 2 3 4
    //4 5 6 1 2 3
    //3 4 5 6 1 2
    //2 3 4 5 6 1

    public int findMin(int[] nums) {
        if (nums[0]<=nums[nums.length-1]) return nums[0];
        int start = 0; int end = nums.length-1;
        while (start < end) {
            int mid = start + (end-start)/2;
            if (mid==0 && nums[mid]<=nums[nums.length-1]) {
                return nums[mid];
            } else if (nums[mid]>nums[end]) {
                start = mid+1;
            } else {
                end = mid;
            }
        }
        return nums[start];
    }
}

File: Data Structures & Algorithms/valid-sudoku/submission-7.py
class Solution:
    def isValidSudoku(self, nums: List[List[str]]) -> bool:
        rows = collections.defaultdict(set) # Auto creates a set if the key does not exist
        cols = collections.defaultdict(set)
        grids = collections.defaultdict(set) # key is a pair: row, grid

        for i in range(0, 9):
            for j in range(0, 9):
                if nums[i][j] in rows[i] and nums[i][j] != '.':
                    return False
                if nums[i][j] in cols[j] and nums[i][j] != '.':
                    return False
                if nums[i][j] in grids[(i//3, j//3)] and nums[i][j] != '.':
                    return False
                if nums[i][j] != '.':
                    rows[i].add(nums[i][j])
                    cols[j].add(nums[i][j])
                    grids[(i//3, j//3)].add(nums[i][j])
        return True